patches/tModLoader/Terraria/ModLoader/Config/ConfigManager.cs

```diff
@@ -356,6 +459,119 @@ public static bool EnumerableEquals(IEnumerable a, IEnumerable b)
+		}
+		return Language.GetText(key).Format(args);
+
+		string FindKeyInScope(string key, string scope)
+		{
+			if (LanguageManager.Instance.Exists(key))
+				return key;
+
+			string[] splitKey = scope.Split(".");
+			for (int j = splitKey.Length - 1; j >= 0; j--) {
+				string partialKey = string.Join(".", splitKey.Take(j + 1));
+				string combinedKey = partialKey + "." + key;
+				if (LanguageManager.Instance.Exists(combinedKey))
+					return combinedKey;
+			}
+
+			return key;
+		}
+	}
+
+	private static T GetAndValidate<T>(MemberInfo memberInfo) where T : ConfigKeyAttribute
+	{
+		var configKeyAttribute = (T)Attribute.GetCustomAttribute(memberInfo, typeof(T));
+		if (configKeyAttribute != null) {
+			if (configKeyAttribute.malformed)
+				throw new ValueNotTranslationKeyException($"{nameof(T)} only accepts localization keys for the 'key' parameter.", errorOnType: memberInfo is Type);
+		}
+		return configKeyAttribute;
+	}
+
+	// Used to determine which key to register, based only on field/property, not class, not necessarily which key to use in UI.
+	internal static string GetConfigKey<T>(MemberInfo memberInfo, string dataName) where T : ConfigKeyAttribute
+	{
+		var configKeyAttribute = GetAndValidate<T>(memberInfo);
+		if (configKeyAttribute != null) {
+			return configKeyAttribute.key;
+		}
+
+		// Autokey: Determine key from the Type the member belongs to.
+		return GetDefaultLocalizationKey(memberInfo, dataName);
+	}
+
+	internal static string GetDefaultLocalizationKey(MemberInfo member, string dataName)
+	{
+		Assembly asm = (member is Type t ? t : member.DeclaringType).Assembly;
+		string groupKey = AssemblyManager.GetAssemblyOwner(asm, out var modName) ? $"Mods.{modName}.Configs" : "Config";
+		string memberKey = member is Type ? member.Name : $"{member.DeclaringType.Name}.{member.Name}";
+		return $"{groupKey}.{memberKey}.{dataName}";
+	}
+
+	internal static string GetLocalizedText<T, TArgs>(PropertyFieldWrapper memberInfo, string dataName, bool defaultToNull) where T : ConfigKeyAttribute where TArgs : ConfigArgsAttribute
+	{
+		// Priority: Provided/AutoKey on member -> Provided/AutoKey on class if member translation is empty string -> member name or null
+		var args = (TArgs)Attribute.GetCustomAttribute(memberInfo.MemberInfo, typeof(TArgs));
+		string configKey = GetConfigKey<T>(memberInfo.MemberInfo, dataName: dataName);
+		if (configKey != null && Language.Exists(configKey)) {
+			string configLocalization = Language.GetTextValue(configKey);
+			if (!string.IsNullOrEmpty(configLocalization))
+				return FormatTextAttribute(configKey, configLocalization, args?.args);
+		}
+
+		var typeConfigKey = (T)Attribute.GetCustomAttribute(memberInfo.Type, typeof(T));
+		if (typeConfigKey != null && Language.Exists(typeConfigKey.key)) {
+			string typeConfigLocalization = Language.GetTextValue(typeConfigKey.key);
+			if (!string.IsNullOrEmpty(typeConfigLocalization))
```

Comment:
`Exists` check is good here, we don't want `Label` to fallback to the type. _but_ the only way to get here would be if the entry in the localization file for `configKey` (`member.Label`) was `""`. In this case, we should just return `""`